File: Cpp/1466-jump-game-v/jump-game-v.cpp
class Solution {
public:

    int n, d;
    int dp[1001];
    vector<int>vec;

    int rec(int i) {
        if(i >= n) return 0;

        if(dp[i] != -1) return dp[i];

        int back = 1, frwd = 1;

        for(int j = i + 1; j <= i + d && j < n; j++) {
            if(vec[i] > vec[j]) 
                frwd = max(frwd , 1 + rec(j));
            else break;
        }

        for(int j = i - 1; (j >= 0) && (j >= i - d); j--) {
            if(vec[i] > vec[j]) 
                back = max(back, 1 + rec(j));
            else break;
        }

        return dp[i] = max(back,frwd);
    }

    int maxJumps(vector<int>& arr, int D) {
        n = arr.size();
        d = D;
        vec = arr;

        memset(dp,-1,sizeof(dp));

        int ans = 1;

        for(int i = 0; i < n; i++) {
            ans = max(ans, rec(i));
        }

        return ans;
    }
};

File: Cpp/1466-jump-game-v/jump-game-v.go
func sol(arr []int, d int, idx int, dp []int) int {
    ans := 0
    if dp[idx] != -1 {
        return dp[idx]
    }
    for i := 1; i <= d; i++ {
        if idx+i < len(arr) {
            if arr[idx+i] >= arr[idx] {
                break
            }
            ans = max(ans, sol(arr, d, idx+i,dp)+1)
        } else {
            break
        }
    }
    for i := 1; i<=d; i++ {
        if idx-i >= 0 {
            if arr[idx-i] >= arr[idx] {
                break
            }
            ans = max(ans, sol(arr, d, idx-i,dp)+1)
        } else {
            break
        }
    }
    dp[idx] = ans
    return dp[idx]
}

func maxJumps(arr []int, d int) int {
    ans := 0
    dp := make([]int, 1001)

    for i := 0; i < 1001; i++ {
        dp[i] = -1
    }

    for i, _ := range arr {
        ans = max(ans, sol(arr, d, i, dp)+1)
    }
    return ans
}